design/baremetal-operator/bmh_non-bootable_iso.md

```diff
@@ -0,0 +1,113 @@
+# Add non-bootable iso attach API to BareMetalHost
+
+## Status
+
+Implementable
+
+## Summary
+
+Add a new field in BareMetalHost so that it is possible to attach a generic
+non-bootable iso image via Ironic, after the host has been provisioned.
+
+## Motivation
+
+In certain scenarios it is desirable to attach a non-bootable iso image
+to a system after provisioning has finished, for example :
+
+A configuration ISO can be used to provide first-boot configuration to a host
+deployed from a standard image in advance. An operator may have a pool of
+powered off hosts with operating system already provisioned. Then, adding
+a host to the cluster boils down to powering it on with a configuration ISO provided.
+
+### Goals
+
+Expose the Ironic API to attach non-bootable iso images via Metal3.
+
+### Non-Goals
+
+Supporting non-ISO images such as a USB stick.
+
+## Proposal
+
+### BMO Proposal
+
+Add a new optional section in the BareMetalHost spec containing details of the
+non-bootable iso image, example :
+
+```yaml
+  spec:
+    dataImage:
+      url: http://1.2.3.4/image.iso
+      credentialsName: dataimage-secret
+      disableCertificateVerification: false
+```
+
+Secret holding the image url credentials, base64 encoded :
+
+```yaml
+---
+apiVersion: v1
+kind: Secret
+metadata:
+  name: dataimage-secret
+type: Opaque
+data:
+  username: dGVzdA==
+  password: cGFzc3dvcmQ=
+```
+
+Note that in this case, the iso image will be attached after the node
+has been provisioned or ExternallyProvisioned.
+
+### Ironic Proposal
+
+The [proposal](https://bugs.launchpad.net/ironic/+bug/2033288) for exposing this functionaility in Ironic has been
+accepted by the upstream Ironic community.
+
+## Design Details
+
+Add an optional section `DataImage` to `BareMetalHostSpec`, with sub-fields
+`Url`, `CredentialsName`, `DisableCertificateVerification` and a secret
+for dataImage with name `CredentialsName`.
+
+The ISO will be attached when the host is in either the `StateProvisioned` or
+the `StateExternallyProvisioned` state before `actionManageSteadyState` is
+called to maintain the State and manage the host power status.
+
+We want to attach the non-bootable ISO without an extra reboot.
```

Comment:
What we need to ensure is that if an operator requests a non-bootable ISO at the same time with a reboot, we first attach the ISO then reboot.